### ISSUE TITLE
Event Module from http://learnyousomeerlang.com/designing-a-concurrent-application

### ISSUE THREAD

**Issue by pwhittin** [suprematic/otplike#24]
The following is my first attempt at implementing the Event Module from [learnyousomeerlang](http://learnyousomeerlang.com/designing-a-concurrent-application) using otplike.

If you could take a look and provide feedback it would really help me make sure 
I’m understanding how to use otplike properly.

```clojure
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; event.clj
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(ns otplike-event-example.event
  (:require [clj-time.coerce :as tc]
            [clj-time.core :as t]
            [clj-time.local :as tl]
            [otplike.process :as op]
            [otplike.proc-util :as opu]))

(defn get-to-go
  "Given a date-time - specified as vectors of integers for year, month and day,
  and hour, minute and second - return the number of milliseconds between now
  and the specified date-time.  If the specified date-time is before now return
  a zero."
  [[year month day] [hour minute second]]
  (let [target-date-time (t/date-time year month day hour minute second 0)
        local-target-date-time (tl/to-local-date-time target-date-time)
        local-target-date-time-millis (tc/to-long local-target-date-time)
        local-now-millis (tc/to-long (tl/local-now))
        to-go (- local-target-date-time-millis local-now-millis)]
    (if (< 0 to-go)
      to-go
      0)))

;; When the specified number of milliseconds 'to-go' elapses send a messge to
;; the specified 'server' process of the form '[:done event-name]'.  While
;; 'to-go' milliseconds has not elapsed, if a message is received from 'server'
;; of the form '[server ref :cancel]' send a message to 'server' of the form
;; '[ref :ok]' and exit the function.
(op/proc-defn
  event-loop
  [server event-name to-go]
  (op/receive!
    [server ref :cancel] (op/! server [ref :ok])
    (after to-go (op/! server [:done event-name]))))

(defn event-start
  "Given a 'server' process, 'event-name', and a time-date represented as two
  integer vectors of the form [year month day] and [hour minute second], start
  an event process named 'event-name' scheduled to send the 'server' a message
  of the form '[:done event-name]' when the time-date is reached.  If the
  time-date is now, or before now, 'server' will receive the message
  immediately.  At any time prior to reaching time-date if 'server' sends a
  message to the started event process of the form '[server ref :cancel]' the
  started event process will send a message to 'server' of the form '[ref :ok]'
  and will terminate execution.  Returns the event process."
  [server event-name year-month-day hour-minute-second]
  (op/spawn
    event-loop
    [server event-name (get-to-go year-month-day hour-minute-second)]))

(defn event-cancel
  "Send a cancel request from 'server' to an 'event' process.  If the event
  process is already canceled it will still work.  This function will not
  return until the event process is canceled.  Returns 'true'."
  [server event]
  (opu/execute-proc!!
    (let [monitor-ref (op/monitor event)]
      (op/! event [server monitor-ref :cancel])
      (op/receive!
        [monitor-ref :ok] (op/demonitor monitor-ref)
        [:DOWN monitor-ref _ _ _] nil)))
  true)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; event_test.clj
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(ns otplike-event-example.event-test
  (:require [clj-time.core :as t]
            [clj-time.local :as tl]
            [clojure.test :refer :all]
            [otplike.process :as op]
            [otplike.proc-util :as opu]
            [otplike-event-example.event :as oe]))

(defn get-now-plus-or-minus
  "Return a time-date instance for now plus or minus the specified 'seconds'."
  [seconds]
  (t/plus (tl/local-now) (t/seconds seconds)))

(defn get-y-m-d
  "Given a time-date instance return a vector of the form '[year month day]'."
  [td]
  [(t/year td) (t/month td) (t/day td)])

(defn get-h-m-s
  "Given a time-date instance return a vector of the form '[hour min sec]'."
  [td]
  [(t/hour td) (t/minute td) (t/second td)])

(defmacro do-receive
  ([done-name] (do-receive &form &env done-name true 0 false))
  ([done-name unique] (do-receive &form &env done-name false unique true))
  ([done-name done-is unique unique-is]
   `(~'op/receive!
      [~unique :ok]
      (~'is ~unique-is "received [ref :ok] message")

      [:done ~done-name]
      (~'is ~done-is "to-go timeout reached")

      :else
      (~'is false "unknown message")

      (~'after 2000 (~'is false "timeout occurred")))))

(deftest get-to-go-test
  (testing "target time after now"
    (let [later (get-now-plus-or-minus 10)]
      (is (< 0 (oe/get-to-go (get-y-m-d later) (get-h-m-s later))))))

  (testing "target time before now"
    (let [later (get-now-plus-or-minus -10)]
      (is (= 0 (oe/get-to-go (get-y-m-d later) (get-h-m-s later)))))))

(deftest event-loop-test
  (testing "to-go time is reached and to-go is greater than zero"
    (opu/execute-proc!!
      (let [name "test"
            server (op/self)]
        (op/spawn oe/event-loop [server name 1000])
        (do-receive name))))

  (testing "to-go time is reached and to-go is zero"
    (opu/execute-proc!!
      (let [name "test"
            server (op/self)]
        (op/spawn oe/event-loop [server name 0])
        (do-receive name))))

  (testing "[ref :ok] message received on cancel with small to-go"
    (opu/execute-proc!!
      (let [name "test"
            unique (gensym)
            server (op/self)
            event (op/spawn oe/event-loop [server name 1000])]
        (op/! event [server unique :cancel])
        (do-receive name unique))))

  (testing "[ref :ok] message received on cancel with very large to-go"
    (opu/execute-proc!!
      (let [years 5
            to-go (* years 365 24 60 60 1000)
            name "test"
            unique (gensym)
            server (op/self)
            event (op/spawn oe/event-loop [server name to-go])]
        (op/! event [server unique :cancel])
        (do-receive name unique)))))

(deftest event-start-test
  (testing "start one event process"
    (opu/execute-proc!!
      (let [server (op/self)
            name "test"
            later (get-now-plus-or-minus 1)
            y-m-d (get-y-m-d later)
            h-m-s (get-h-m-s later)]
        (oe/event-start server name y-m-d h-m-s)
        (do-receive name))))

  (testing "start multiple event processes"
    (opu/execute-proc!!
      (let [server (op/self)
            names #{"test-1" "test-2" "test-3"}
            later (get-now-plus-or-minus 1)
            y-m-d (get-y-m-d later)
            h-m-s (get-h-m-s later)]
        (doseq [name names]
          (oe/event-start server name y-m-d h-m-s))
        (loop [done-names #{}]
          (op/receive!
            [:done n]
            (let [names-so-far (conj done-names n)]
              (if (= names-so-far names)
                (is true "all events done")
                (recur names-so-far)))

            :else
            (is false "unknown message")

            (after 2000 (is false "timeout occurred"))))))))

(deftest event-cancel-test
  (testing "cancel running event process"
    (opu/execute-proc!!
      (let [server (op/self)
            name "test"
            later (get-now-plus-or-minus 1)
            y-m-d (get-y-m-d later)
            h-m-s (get-h-m-s later)
            event (oe/event-start server name y-m-d h-m-s)]
        (is (oe/event-cancel server event)))))

  (testing "cancel already canceled event process"
    (opu/execute-proc!!
      (let [server (op/self)
            unique (gensym)
            name "test"
            later (get-now-plus-or-minus 1)
            y-m-d (get-y-m-d later)
            h-m-s (get-h-m-s later)
            event (oe/event-start server name y-m-d h-m-s)]
        (op/! event [server unique :cancel])
        (do-receive name unique)
        (is (oe/event-cancel server event))))))
```

**Comment by pwhittin** [comment]
Within event_test.clj you'll notice that 'do-receive' is a macro.  I had to make it a macro because it did not work as a function.  It compiles, but when the tests are executed I get the following error associated with the call to 'receive!':
```
java.lang.AssertionError: Assert failed: alts! used not in (go ...) block
```
The following code is the implementation I tested of 'do-receive' as a function:
```
(defn do-receive
  ([done-name] (do-receive done-name true 0 false))
  ([done-name unique] (do-receive done-name false unique true))
  ([done-name done-is unique unique-is]
   (op/receive!
     [unique :ok]
     (is unique-is "received [ref :ok] message")

     [:done done-name]
     (is done-is "to-go timeout reached")

     :else
     (is false "unknown message")

     (after 2000 (is false "timeout occurred")))))
```

The odd thing is that the following similar code executes perfectly in a newly started 'lein repl':
```
(require '[otplike.process :as process :refer [! pid->str self]])
(require '[clojure.core.async :as async])

(defn my-receive
  [msg1 msg2 timeout self]
  (process/receive!
    msg (printf "[%s] %s %s%n" self msg1 msg)
    (after timeout
           (printf "[%s] %s %s%n" self msg2 timeout))))

(process/proc-defn
  await-message
  [timeout]
  (my-receive "receive" "timeout" timeout (process/self)))

(defn run []
  (let [pid0 (process/spawn await-message [1000])]
    (println "[main] Started [" pid0 "]")
    (let [pid1 (process/spawn await-message [(async/timeout 1000)])]
      (println "[main] Started [" pid1 "]")
      (let [pid2 (process/spawn await-message [:infinity])]
        (println "[main] Started [" pid2 "]")
        (println "[main] Waiting for 2 seconds")
        (Thread/sleep 2000)
        (println "[main] Sending :stop to [" pid2 "]")
        (! pid2 :stop)
        (println "[main] Waiting for 100mS")
        (Thread/sleep 100)
        (println "[main] Done")))))

(run)
```

**Comment by sskorokhodov** [comment]
Could you please make a new issue for the second comment? It's rather important question on its own. I'd better answer it separately if you don't mind.

**Comment by sskorokhodov** [comment]
Finally, I have time to answer this question :)

The short answer is - you did everyting right. The rest is about limitations of otplike as of `0.3.x-alpha`.

---

The only (but very interesting) part of your code I want to talk about is this function:

```clojure
(defn event-cancel
  "Send a cancel request from 'server' to an 'event' process.  If the event
  process is already canceled it will still work.  This function will not
  return until the event process is canceled.  Returns 'true'."
  [server event]
  (opu/execute-proc!!
    (let [monitor-ref (op/monitor event)]
      (op/! event [server monitor-ref :cancel])
      (op/receive!
        [monitor-ref :ok] (op/demonitor monitor-ref)
        [:DOWN monitor-ref _ _ _] nil)))
  true)
```

The good thing about it - it works as expected. **And this is almost the only way to do it now.**

What is not so good:

1. Each call to `event-cancel` creates a new process. Which is not as lightweight as I wish it was.
2. Each call blocks the "real" thread. In case of multiple simultaneous calls it can lead to situation when each thread is blocked waiting for result and no thread awailable to compute and deliver the result.

The code below shows how gen-server/call is implemented in `0.3.x-alpha`.

```clojure
(defn call* [server message timeout-ms]
  (process/async
    (let [reply-to (async/chan)
          timeout (if (= :infinity timeout-ms)
                    (async/chan)
                    (async/timeout timeout-ms))]
      (if-not (! server [::call reply-to message])
        [:error :noproc]
        (match (async/alts! [reply-to timeout])
          [[::terminated reason] reply-to] [:error reason]
          [[::reply value] reply-to] [:ok value]
          [nil timeout] [:error :timeout])))))
```

As you can see, we use channel to get the answer back.

The problem with this code is that you can not monitor the server while waiting for the answer. Checking the result of `!` ensures that the message has been delivered to the server's inbox. Some code on the server side ensures that reply will be sent back even if the server crashed while processing the call. But if the server exits after the message has been placed into its inbox but before it started processing the call, reply will be never sent back.

It is our compromise between non-blocking "fast" call in common case and possible delay for the time of the timeout. And that's why as of now it's not recommended to use gen-server/call without timeout.

I hope to implement `process/selective-receive!` and "flushing" `process/demonitor` in `0.4.0-alpha` to make possible fast and reliable calls.

Also there is an additional tinkering with the timeout value, which would have been unnecessary with selective receive.

---

I don't have a lot to say about the tests. It seems like they do what they are supposed to do :)

**Comment by pwhittin** [comment]
As you suggested in issue 26 (https://github.com/suprematic/otplike/issues/26), using "receive!!" allowed me to make do-receive just a regular function:
``` clojure
(defn do-receive
  ([done-name] (do-receive done-name true 0 false))
  ([done-name unique] (do-receive done-name false unique true))
  ([done-name done-is unique unique-is]
   (op/receive!!                  ;; <<<<<===== Here is the change
     [unique :ok]
     (is unique-is "received [ref :ok] message")

     [:done done-name]
     (is done-is "to-go timeout reached")

     :else
     (is false "unknown message")

     (after 2000 (is false "timeout occurred")))))
```
